### PR TITLE
GEOMESA-1369 Extra classpaths for mapreduce

### PR DESCRIPTION
**Pull request by jahhulbert-ccri**
* Load Converters and Transformers in GEOMESA_EXTRA_CLASSPATH for MapReduce
* Make extra classpaths load for libjars in MapReduce
* Make extra classpaths like Java classpath
* Don't recurse into directories on EXTRA classpaths

### CODE REVIEW

File: geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/accumulo/tools/AccumuloRunner.scala [hunk 1]
```diff
@@ -93,7 +93,9 @@ object AccumuloRunner extends Runner {
       if (p.zookeepers == null) {
         p.zookeepers = zookeepers
       }
-      if (p.instance == null) {
+
+      // Attempt to look up the instance ONLY if zookeepers is set and we are not in mock mode
+      if (p.instance == null && p.zookeepers != null && !p.mock ) {
```
Comment (aheyne) [inline geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/accumulo/tools/AccumuloRunner.scala]: Just for reference, this is a similar change to #1284

Comment (jahhulbert-ccri) [inline geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/accumulo/tools/AccumuloRunner.scala]: @aheyne cool I'll rebase on that and take a look at that PR

File: geomesa-accumulo/geomesa-accumulo-tools/bin/geomesa [hunk 2]
```diff
@@ -56,9 +56,9 @@ fi
 # include geomesa first so that the correct log4j.properties is picked up
 GEOMESA_CP="${GEOMESA_CONF_DIR}:$(findJars $GEOMESA_LIB):${ACCUMULO_CONF_DIR}:${HADOOP_CONF_DIR}"
 
-# Add user defined directories to the classpath
-if [[ -z "$GEOMESA_EXTRA_CLASSPATHS"  ]]; then
-  GEOMESA_CP="${GEOMESA_CP}:$(findJars $GEOMESA_EXTRA_CLASSPATHS true)"
+# Add user defined directories to the classpath using java classpath syntax
+if [[ -n "${GEOMESA_EXTRA_CLASSPATH}" ]]; then
+  GEOMESA_CP="${GEOMESA_CP}:${GEOMESA_EXTRA_CLASSPATH}"
```
Comment (jahhulbert-ccri) [inline geomesa-accumulo/geomesa-accumulo-tools/bin/geomesa]: Excellent point. I think prepend is appropriate

File: geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/accumulo/tools/ingest/AccumuloIngestCommand.scala [hunk 3]
```diff
@@ -24,9 +25,11 @@ class AccumuloIngestCommand extends IngestCommand[AccumuloDataStore] with Accumu
   override val libjarsFile: String = "org/locationtech/geomesa/accumulo/tools/ingest-libjars.list"
 
   override def libjarsPaths: Iterator[() => Seq[File]] = Iterator(
-    () => ClassPathUtils.getJarsFromEnvironment("GEOMESA_ACCUMULO_HOME"),
-    () => ClassPathUtils.getJarsFromEnvironment("GEOMESA_HOME"),
-    () => ClassPathUtils.getJarsFromEnvironment("ACCUMULO_HOME"),
+    () => {
+      val dirs = AccumuloClassPathAdditions.Directories
```
Comment (jahhulbert-ccri) [inline geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/accumulo/tools/ingest/AccumuloIngestCommand.scala]: :) yeah I can do that instead

File: docs/user/accumulo/install.rst [hunk 4]
```diff
@@ -168,9 +168,10 @@ Update and re-source your ``~/.bashrc`` file to include the ``$GEOMESA_HOME`` an
     ``geomesa`` will read the ``$ACCUMULO_HOME`` and ``$HADOOP_HOME`` environment variables to load the
     appropriate JAR files for Hadoop, Accumulo, Zookeeper, and Thrift. If possible, we recommend
     installing the tools on the Accumulo master server, as you may also need various configuration
-    files from Hadoop/Accumulo in order to run certain commands. In addition ``geomesa`` will pull any
-    additional jars from the ``$GEOMESA_EXTRA_CLASSPATHS`` environment variable into the class path.
-    Use the ``geomesa classpath`` command in order to see what JARs are being used.
+    files from Hadoop/Accumulo in order to run certain commands. In addition ``geomesa`` will append the environmental
```
Comment (jahhulbert-ccri) [inline docs/user/accumulo/install.rst]: will do

File: geomesa-accumulo/geomesa-accumulo-tools/pom.xml [hunk 5]
```diff
@@ -161,6 +161,96 @@
                     <failOnUnableToExtractRepoInfo>false</failOnUnableToExtractRepoInfo>
                 </configuration>
             </plugin>
+            <plugin>
```
Comment (jahhulbert-ccri) [inline geomesa-accumulo/geomesa-accumulo-tools/pom.xml]: yeah

**Comment by jahhulbert-ccri** [comment]
still working fyi

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala [hunk 6]
```diff
@@ -136,4 +136,20 @@ object ClassPathUtils extends LazyLogging {
   def cleanClassPathURL(url: String): String =
     URLDecoder.decode(url, "UTF-8").replace("file:", "").replace("!", "")
 
+  def loadClassPathFromEnv(prop: String): Seq[File] = {
```
Comment (jahhulbert-ccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala]: I am attempting to replicate the HADOOP_CLASSPATHS code...which I think is:
1. colon (```:```) separated entries
2. if it ends with ```/*``` then expand out and fetch just jars...seems that java operates that way
3. take specific files if listed.
4. subdirs are not recursively searched but all files in a dir are fetched (not just jars)

**Comment by jahhulbert-ccri** [comment]
@mdzimmerman take another look. I'm also going to get Brent to check it out

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala (hunk 6 above)
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala]: makes sense - can we add that as javadoc on the method?

File: geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/JobUtils.scala [hunk 7]
```diff
@@ -24,7 +24,10 @@ object JobUtils extends LazyLogging {
    * @param libJars jar prefixes to load
    */
   def setLibJars(conf: Configuration, libJars: Seq[String], searchPath: Iterator[() => Seq[File]]): Unit = {
-    val paths = ClassPathUtils.findJars(libJars, searchPath).map(f => "file:///" + f.getAbsolutePath)
+    val extra = ClassPathUtils.loadClassPathFromEnv("GEOMESA_EXTRA_CLASSPATHS")
```
Comment (elahrvivaz) [inline geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/JobUtils.scala]: will `libjars` work with non-jar files? for some reason i thought that they actually had to end with .jar

Comment (jahhulbert-ccri) [inline geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/JobUtils.scala]: i think so. i was able to send a converter conf in i think

**Comment by jahhulbert-ccri** [comment]
@elahrvivaz i'm going to run through some tests really quick:
1. Provide a custom converter on the command line in a file (should work in distributed s3 ingest)
2. Provide a jar file with custom transformers (should work in distributed s3 ingest)
3. Provide a convert script in javascript  (should work in distributed s3 ingest)

File: geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/JobUtils.scala (hunk 7 above)
Comment (jahhulbert-ccri) [inline geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/JobUtils.scala]: i don't know if libjars on command line will but it seemed to work...i'll verify

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala (hunk 6 above)
Comment (jahhulbert-ccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/classpath/ClassPathUtils.scala]: prob should add some doc somewhere too one day :)

File: geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/JobUtils.scala (hunk 7 above)
Comment (jahhulbert-ccri) [inline geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/JobUtils.scala]: seemed to work for files just fine.

Comment (jahhulbert-ccri) [inline geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/JobUtils.scala]: more testing might be good!

**Comment by jahhulbert-ccri** [comment]
@elahrvivaz  i changed these to a colon separated list...i can revert to commas...any thoughts?

**Comment by elahrvivaz** [comment]
@jahhulbert-ccri changed what from commas?

**Comment by jahhulbert-ccri** [comment]
the contents of the variable ```geomesa.convert.scripts.path```

**Comment by elahrvivaz** [comment]
works for me

**Comment by jahhulbert-ccri** [comment]
cool. well in that case i guess its good. I will relook at other classpath fun later but I think that clears up the usecase for distributed ingest with custom stuff.

**Comment by elahrvivaz** [comment]
@jahhulbert-ccri ready for merge?

**Comment by jahhulbert-ccri** [comment]
:shipit:

**Pull request merged**